### PR TITLE
Move `Pruning` into `physical-optimizer` crate

### PR DESCRIPTION
**Pull request by irenjj**
## Which issue does this PR close?

part of https://github.com/apache/datafusion/issues/11502.

### CODE REVIEW

File: datafusion/physical-optimizer/Cargo.toml [hunk 1]
```diff
@@ -33,13 +33,17 @@ workspace = true
 
 [dependencies]
 arrow = { workspace = true }
+arrow-array = { workspace = true }
 arrow-schema = { workspace = true }
 datafusion-common = { workspace = true, default-features = true }
 datafusion-execution = { workspace = true }
 datafusion-expr-common = { workspace = true, default-features = true }
+datafusion-expr = { workspace = true, default-features = true }
+datafusion-functions-nested = { workspace = true }
```
Comment (alamb) [inline datafusion/physical-optimizer/Cargo.toml:42]: Can you please move `datafusion-functions-nested` to `dev-dependencies`?

I also think we can avoid the dependency on `datafusion-expr` by changing 

```rust
use datafusion_expr::Operator;
```

to 

```rust
use datafusion_expr_common::operator::Operator;
```

File: datafusion/physical-optimizer/Cargo.toml [hunk 2]
```diff
@@ -33,13 +33,17 @@ workspace = true
 
 [dependencies]
 arrow = { workspace = true }
+arrow-array = { workspace = true }
```
Comment (alamb) [inline datafusion/physical-optimizer/Cargo.toml:36]: If possible, can you avoid this explicit dependency? Whatever is used in arrow-array should be available in `arrow` and then we can keep the dependency chain simpler

**Review by alamb — COMMENTED** [review]
Thank you @irenjj  -- this looks very nice!

I think we need to clean up some of the dependencies, but it is really close

Thank you!

File: datafusion/physical-optimizer/src/pruning.rs [hunk 3]
```diff
@@ -22,12 +22,10 @@
 use std::collections::HashSet;
```
Comment (irenjj) [inline datafusion/physical-optimizer/src/pruning.rs:22]: `datafusion-expr` dependency is needed by `//! [`Expr`]: crate::prelude::Expr`, should we remove the comment? @alamb

Comment (alamb) [reply datafusion/physical-optimizer/src/pruning.rs:22]: In other places, we have changed the link to docs.rs directly. So in this case, change the link to

https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.Expr.html

File: datafusion/physical-optimizer/Cargo.toml [hunk 4]
```diff
@@ -37,11 +37,15 @@ arrow-schema = { workspace = true }
...
 datafusion-expr-common = { workspace = true, default-features = true }
+datafusion-optimizer = { workspace = true }
 datafusion-physical-expr = { workspace = true }
 datafusion-physical-plan = { workspace = true }
 itertools = { workspace = true }
+log = { workspace = true }
 recursive = { workspace = true }
 
 [dev-dependencies]
+datafusion-expr = { workspace = true }
```
Comment (irenjj) [inline datafusion/physical-optimizer/Cargo.toml:48]: needed in test.

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @irenjj

**Comment by jayzhan211** [comment]
Thanks @irenjj @alamb

**Pull request merged**